crates/autopilot/src/fok_limit_orders/quoter.rs

```diff
@@ -191,35 +191,54 @@ async fn orders_with_sufficient_balance(
     balance_fetcher: &dyn BalanceFetching,
     mut orders: Vec<OrderQuotingData>,
 ) -> Vec<OrderQuotingData> {
+    // Note that we skip balance checks for orders with pre-interactions. This
+    // notably includes EthFlow orders, as the WETH for the trade will get
+    // deposited as part of a pre-interaction and might not be available when
+    // checking whether or not we should be. This exception is also needed for
+    // user orders with custom pre-interactions (for example, an order with a
+    // EIP-2612 `permit` pre-interaction to set an allowance).
+    let skip_balance_check = |order: &OrderQuotingData| order.pre_interactions > 0;
+
     let queries = orders
         .iter()
-        .filter(|order| order.pre_interactions <= 0)
+        .filter(|order| !skip_balance_check(order))
         .map(query_from)
         .unique()
         .collect_vec();
     let balances = balance_fetcher.get_balances(&queries).await;
-    let balances: HashMap<_, _> = queries
-        .iter()
-        .zip(balances.into_iter())
-        .filter_map(|(query, result)| Some((query, result.ok()?)))
-        .collect();
+    let balances = {
```

Comment:
Mostly because I dislike side-effects in `Iterator` methods like `filter` and `map`.

That being said, I break my own rule and add them to the `retain` closure... so will change it to be consistent.